## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=['EIN', 'NAME'])

In [3]:
# Determine the number of unique values in each column.
unique_values = application_df.nunique()

# Display the unique values count for each column
unique_values

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts to identify and replace with "Other"
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
print(application_type_counts)





T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64


In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
print(application_type_counts)

# Choose a cutoff value and create a list of application types to be replaced
cutoff_value = 528
application_types_to_replace = list(application_type_counts[application_type_counts < cutoff_value].index) # поменял <= на <, чтобы соответствовать ожидаемому результату



# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure replacement was successful
application_df['APPLICATION_TYPE'].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64


T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts to identify and replace with "Other"
classification_counts = application_df['CLASSIFICATION'].value_counts()
print(classification_counts)


C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64


In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1
# Look at CLASSIFICATION value counts for binning
classification_counts = application_df['CLASSIFICATION'].value_counts()
print(classification_counts[classification_counts > 1])



C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64


In [8]:
# Choose a cutoff value and create a list of classifications to be replaced
cutoff_value = 1883
classifications_to_replace = list(classification_counts[classification_counts < cutoff_value].index)


# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure replacement was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [9]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df = pd.get_dummies(application_df)

In [10]:
# Split our preprocessed data into our features and target arrays
target_column = 'IS_SUCCESSFUL'
y = application_df[target_column].values
X = application_df.drop(columns=[target_column]).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [11]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [12]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_features = X_train.shape[1]

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=80, activation='relu', input_dim=input_features, dtype='float32'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=30, activation='relu', dtype='float32'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid', dtype='float32'))

# Check the structure of the model
nn.summary()

c:\Users\13465\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 80)             │         3,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 30)             │         2,430 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            31 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 5,981 (23.36 KB)

 Trainable params: 5,981 (23.36 KB)

 Non-trainable params: 0 (0.00 B)

In [13]:
# Compile the model
nn.compile(optimizer='adam', 
           loss='binary_crossentropy', 
           metrics=['accuracy'])

# Check the structure of the model
nn.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 80)             │         3,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 30)             │         2,430 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            31 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 5,981 (23.36 KB)

 Trainable params: 5,981 (23.36 KB)

 Non-trainable params: 0 (0.00 B)

In [14]:
# Train the model
history = nn.fit(X_train_scaled, y_train, epochs=100, batch_size=32, validation_data=(X_test_scaled, y_test))

Epoch 1/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.7056 - loss: 0.5874 - val_accuracy: 0.7236 - val_loss: 0.5650
Epoch 2/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7301 - loss: 0.5551 - val_accuracy: 0.7213 - val_loss: 0.5616
Epoch 3/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.7314 - loss: 0.5472 - val_accuracy: 0.7242 - val_loss: 0.5588
Epoch 4/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.7369 - loss: 0.5441 - val_accuracy: 0.7248 - val_loss: 0.5570
Epoch 5/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.7315 - loss: 0.5485 - val_accuracy: 0.7232 - val_loss: 0.5605
Epoch 6/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7303 - loss: 0.5482 - val_accuracy: 0.7277 - val_loss: 0.5587
Epoch 7/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 1s 930us/step - accuracy: 0.7290 - loss: 0.5504 - val_accuracy: 0.7273 - val_loss: 0.5564
Epoch 8/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7358 - loss: 0.5416 - val_ac

In [15]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - 2ms/step - accuracy: 0.7271 - loss: 0.5690
Loss: 0.569021463394165, Accuracy: 0.7271137237548828


I tried different batch sizes ranging from 1 to 64. Since the changes in accuracy and loss with different batch sizes were not significant, I chose a larger batch size for the convenience of training and numerical stability.

In [16]:
# Export our model to HDF5 file
nn.save('AlphabetSoupCharity.h5')

In the first version of the model, we use the `adam` optimizer as the most advanced optimization method because it utilizes the first and second derivatives to avoid local minima. This helps prevent the model from getting stuck in a local minimum, ensuring it is well-trained.

In [17]:
# We use the SGD optimizer - Stochastic Gradient Descent.

# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_features = X_train.shape[1]

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=80, activation='relu', input_dim=input_features, dtype='float32'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=30, activation='relu', dtype='float32'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid', dtype='float32'))

# Check the structure of the model
nn.summary()

# Compile the model
nn.compile(optimizer='sgd', 
           loss='binary_crossentropy', 
           metrics=['accuracy'])

# Check the structure of the model
nn.summary()

history = nn.fit(X_train_scaled, y_train, epochs=100, batch_size=32, validation_data=(X_test_scaled, y_test))

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 80)             │         3,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 30)             │         2,430 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 1)              │            31 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 5,981 (23.36 KB)

 Trainable params: 5,981 (23.36 KB)

 Non-trainable params: 0 (0.00 B)

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 80)             │         3,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 30)             │         2,430 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 1)              │            31 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 5,981 (23.36 KB)

 Trainable params: 5,981 (23.36 KB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.6023 - loss: 0.6562 - val_accuracy: 0.7141 - val_loss: 0.5825
Epoch 2/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7222 - loss: 0.5682 - val_accuracy: 0.7211 - val_loss: 0.5704
Epoch 3/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7274 - loss: 0.5596 - val_accuracy: 0.7229 - val_loss: 0.5668
Epoch 4/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7322 - loss: 0.5560 - val_accuracy: 0.7219 - val_loss: 0.5655
Epoch 5/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7294 - loss: 0.5532 - val_accuracy: 0.7252 - val_loss: 0.5633
Epoch 6/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7245 - loss: 0.5579 - val_accuracy: 0.7224 - val_loss: 0.5626
Epoch 7/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7309 - loss: 0.5509 - val_accuracy: 0.7233 - val_loss: 0.5620
Epoch 8/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7337 - loss: 0.5492 - val_accu

In [18]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - 581us/step - accuracy: 0.7242 - loss: 0.5572
Loss: 0.55715012550354, Accuracy: 0.7241982221603394


Aligning with classical expectations, `adam` performs slightly better than `sgd`, with an `Accuracy` of 0.724781334400177 compared to `sgd`'s `Accuracy` of 0.72434401512146, and a `Loss` of 0.5697312951087952 compared to sgd's `Loss` of 0.5590302348136902. 

In [19]:
# We use the rmsprop optimizer 

# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_features = X_train.shape[1]

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=80, activation='relu', input_dim=input_features, dtype='float32'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=30, activation='relu', dtype='float32'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid', dtype='float32'))

# Check the structure of the model
nn.summary()

# Compile the model
nn.compile(optimizer='rmsprop', 
           loss='binary_crossentropy', 
           metrics=['accuracy'])

# Check the structure of the model
nn.summary()

history = nn.fit(X_train_scaled, y_train, epochs=100, batch_size=32, validation_data=(X_test_scaled, y_test))

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_6 (Dense)                 │ (None, 80)             │         3,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 30)             │         2,430 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 1)              │            31 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 5,981 (23.36 KB)

 Trainable params: 5,981 (23.36 KB)

 Non-trainable params: 0 (0.00 B)

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_6 (Dense)                 │ (None, 80)             │         3,520 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 30)             │         2,430 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 1)              │            31 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 5,981 (23.36 KB)

 Trainable params: 5,981 (23.36 KB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 1s 943us/step - accuracy: 0.6925 - loss: 0.5935 - val_accuracy: 0.7261 - val_loss: 0.5660
Epoch 2/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 1s 820us/step - accuracy: 0.7307 - loss: 0.5566 - val_accuracy: 0.7195 - val_loss: 0.5627
Epoch 3/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 1s 804us/step - accuracy: 0.7306 - loss: 0.5524 - val_accuracy: 0.7248 - val_loss: 0.5599
Epoch 4/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 1s 778us/step - accuracy: 0.7325 - loss: 0.5472 - val_accuracy: 0.7258 - val_loss: 0.5608
Epoch 5/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 1s 806us/step - accuracy: 0.7315 - loss: 0.5508 - val_accuracy: 0.7232 - val_loss: 0.5590
Epoch 6/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 1s 781us/step - accuracy: 0.7338 - loss: 0.5453 - val_accuracy: 0.7226 - val_loss: 0.5620
Epoch 7/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 1s 834us/step - accuracy: 0.7345 - loss: 0.5489 - val_accuracy: 0.7270 - val_loss: 0.5628
Epoch 8/100
858/858 ━━━━━━━━━━━━━━━━━━━━ 1s 799us/step - accuracy: 0.7350 - loss: 0

In [20]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - 536us/step - accuracy: 0.7267 - loss: 0.5613
Loss: 0.5613489151000977, Accuracy: 0.7266764044761658


`rmsprop` shows the best performance

`rmsprop` turned out to be the most efficient optimizer based on the results of three training sessions:

1. `rmsprop`
2. `adam`
3. `sgd`
The main difference between `rmsprop` and `sgd` is that rmsprop uses adaptive learning rates for each weight. `adam` utilizes a larger number of mathematical criteria to find the global minimum and will likely perform best in this task with finer tuning of each parameter.

Let's try to work with the data and with the layers of the model to improve the result.



In [21]:
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [22]:
# Use another variant for model optimization: retain the "Name" column in the data,
# as it contains important information, handle outliers in the data more accurately, and add more hidden layers to the neural network model.
application_df = application_df.drop(columns=['EIN'])

In [23]:
# Determine the number of unique values in each column.
unique_values = application_df.nunique()

# Display the unique values count for each column
unique_values

NAME                      19568
APPLICATION_TYPE             17
AFFILIATION                   6
CLASSIFICATION               71
USE_CASE                      5
ORGANIZATION                  4
STATUS                        2
INCOME_AMT                    9
SPECIAL_CONSIDERATIONS        2
ASK_AMT                    8747
IS_SUCCESSFUL                 2
dtype: int64

In [24]:
# Look at APPLICATION_TYPE value counts to identify and replace with "Other"
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
print(application_type_counts)



T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64


In [25]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
print(application_type_counts)

# Choose a cutoff value and create a list of application types to be replaced
cutoff_value = 500
application_types_to_replace = list(application_type_counts[application_type_counts <= cutoff_value].index)



# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure replacement was successful
application_df['APPLICATION_TYPE'].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64


T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [26]:
# Look at CLASSIFICATION value counts to identify and replace with "Other"
classification_counts = application_df['CLASSIFICATION'].value_counts()
print(classification_counts)


C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64


In [27]:
# You may find it helpful to look at CLASSIFICATION value counts >1
# Look at CLASSIFICATION value counts for binning
classification_counts = application_df['CLASSIFICATION'].value_counts()
print(classification_counts[classification_counts > 1])

# Choose a cutoff value and create a list of classifications to be replaced
cutoff_value = 100
classification_types_to_replace = list(classification_counts[classification_counts < cutoff_value].index)

# Replace in dataframe
for cls in classification_types_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls, "Other")

# Check to make sure replacement was successful
print(application_df['CLASSIFICATION'].value_counts())

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
Other      669
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: CLASSIFICATION, dtype: int64


In [28]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classification_counts = application_df['CLASSIFICATION'].value_counts()
print(classification_counts[classification_counts > 1])

# Choose a cutoff value and create a list of classifications to be replaced
cutoff_value = 1000
classifications_to_replace = list(classification_counts[classification_counts < cutoff_value].index)


# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure replacement was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
Other      669
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: CLASSIFICATION, dtype: int64


C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [29]:
name_count = application_df['NAME'].value_counts()
name_count

PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: NAME, Length: 19568, dtype: int64

In [30]:
names_to_replace = list(name_count[name_count< 100].index)

for name in names_to_replace:
    application_df['NAME'] = application_df['NAME'].replace(
        name, "Other")

application_df['NAME'].value_counts()

Other                                                                 25987
PARENT BOOSTER USA INC                                                 1260
TOPS CLUB INC                                                           765
UNITED STATES BOWLING CONGRESS INC                                      700
WASHINGTON STATE UNIVERSITY                                             492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                         408
PTA TEXAS CONGRESS                                                      368
SOROPTIMIST INTERNATIONAL OF THE AMERICAS INC                           331
ALPHA PHI SIGMA                                                         313
TOASTMASTERS INTERNATIONAL                                              293
MOST WORSHIPFUL STRINGER FREE AND ACCEPTED MASONS                       287
LITTLE LEAGUE BASEBALL INC                                              277
INTERNATIONAL ASSOCIATION OF LIONS CLUBS                                266
MOMS CLUB   

In [31]:
app_dummied = pd.get_dummies(application_df)
app_dummied.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  NAME_ALPHA PHI SIGMA  \
0       1     5000              1                     0   
1       1   108590              1                     0   
2       1     5000              0                     0   
3       1     6692              1                     0   
4       1   142590              1                     0   

   NAME_AMATEUR ATHLETIC UNION OF THE UNITED STATES INC  \
0                                                  0      
1                                                  0      
2                                                  0      
3                                                  0      
4                                                  0      

   NAME_AMERICAN ASSOCIATION OF UNIVERSITY WOMEN  NAME_CIVITAN INTERNATIONAL  \
0                                              0                           0   
1                                              0                           0   
2                                              0                           0   
3                                              0                           0   
4                                              0                           0   

   NAME_DEMOLAY INTERNATIONAL  \
0                           0   
1                           0   
2                           0   
3                           0   
4                           0   

   NAME_FARMERS EDUCATIONAL AND COOPERATIVE UNION OF AMERICA  \
0                                                  0           
1                                                  0           
2                                                  0           
3                                                  0           
4                                                  0           

   NAME_HABITAT FOR HUMANITY INTERNATIONAL INC  ...  INCOME_AMT_1-9999  \
0                                            0  ...                  0   
1                                            0  ...                  1   
2                                            0  ...                  0   
3                                            0  ...                  0   
4                                            0  ...                  0   

   INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                       0                         0                   0   
1                       0                         0                   0   
2                       0                         0                   0   
3                       1                         0                   0   
4                       0                         1                   0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0                 0                       0                0   
1                 0                       0                0   
2                 0                       0                0   
3                 0                       0                0   
4                 0                       0                0   

   INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
0                  0                         1                         0  
1                  0                         1                         0  
2                  0                         1                         0  
3                  0                         1                         0  
4                  0                         1                         0  

[5 rows x 75 columns]

In [32]:
X = app_dummied.drop("IS_SUCCESSFUL", axis=1)
y = app_dummied.IS_SUCCESSFUL
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [33]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [34]:
X_train.head()

STATUS   ASK_AMT  NAME_ALPHA PHI SIGMA  \
30531       1      5000                     0   
22775       1      5000                     0   
19057       1      5000                     0   
27536       1      5000                     0   
21331       1  11121098                     0   

       NAME_AMATEUR ATHLETIC UNION OF THE UNITED STATES INC  \
30531                                                  0      
22775                                                  0      
19057                                                  0      
27536                                                  0      
21331                                                  0      

       NAME_AMERICAN ASSOCIATION OF UNIVERSITY WOMEN  \
30531                                              0   
22775                                              0   
19057                                              0   
27536                                              0   
21331                                              0   

       NAME_CIVITAN INTERNATIONAL  NAME_DEMOLAY INTERNATIONAL  \
30531                           0                           0   
22775                           0                           0   
19057                           0                           0   
27536                           0                           0   
21331                           0                           0   

       NAME_FARMERS EDUCATIONAL AND COOPERATIVE UNION OF AMERICA  \
30531                                                  0           
22775                                                  0           
19057                                                  0           
27536                                                  0           
21331                                                  0           

       NAME_HABITAT FOR HUMANITY INTERNATIONAL INC  \
30531                                            0   
22775                                            0   
19057                                            0   
27536                                            0   
21331                                            0   

       NAME_HONOR SOCIETY OF PHI KAPPA PHI  ...  INCOME_AMT_1-9999  \
30531                                    0  ...                  0   
22775                                    0  ...                  0   
19057                                    0  ...                  0   
27536                                    0  ...                  0   
21331                                    0  ...                  0   

       INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
30531                       0                         0                   0   
22775                       0                         0                   0   
19057                       0                         0                   0   
27536                       0                         0                   0   
21331                       0                         0                   1   

       INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
30531                 0                       0                0   
22775                 0                       0                0   
19057                 0                       0                0   
27536                 0                       0                0   
21331                 0                       0                0   

       INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
30531                  0                         1                         0  
22775                  0                         1                         0  
19057                  0                         1                         0  
27536                  0                         1                         0  
21331                  0                         1                         0  

[5 rows x 74 columns]

In [35]:
number_input_features = len(X_train_scaled[0])

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=7, input_dim=number_input_features, activation='relu'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=14, activation='relu'))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=21, activation='tanh'))

# Fourth hidden layer
nn.add(tf.keras.layers.Dense(units=21, activation='linear'))

# Fifth hidden layer
nn.add(tf.keras.layers.Dense(units=28, activation='relu'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

c:\Users\13465\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_9 (Dense)                 │ (None, 7)              │           525 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_10 (Dense)                │ (None, 14)             │           112 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_11 (Dense)                │ (None, 21)             │           315 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_12 (Dense)                │ (None, 21)             │           462 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_13 (Dense)                │ (None, 28)             │           616 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_14 (Dense)                │ (None, 1)              │            29 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,059 (8.04 KB)

 Trainable params: 2,059 (8.04 KB)

 Non-trainable params: 0 (0.00 B)

In [36]:
nn.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics=['accuracy'])

In [37]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,validation_split=0.15, epochs=200)

Epoch 1/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.6701 - loss: 0.5912 - val_accuracy: 0.7396 - val_loss: 0.5209
Epoch 2/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7550 - loss: 0.5013 - val_accuracy: 0.7432 - val_loss: 0.5073
Epoch 3/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7555 - loss: 0.4994 - val_accuracy: 0.7455 - val_loss: 0.5007
Epoch 4/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7599 - loss: 0.4964 - val_accuracy: 0.7476 - val_loss: 0.4970
Epoch 5/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7535 - loss: 0.4975 - val_accuracy: 0.7484 - val_loss: 0.4987
Epoch 6/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7619 - loss: 0.4893 - val_accuracy: 0.7507 - val_loss: 0.4944
Epoch 7/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7593 - loss: 0.4878 - val_accuracy: 0.7499 - val_loss: 0.4964
Epoch 8/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 997us/step - accuracy: 0.7597 - loss: 0.4881 - val_ac

In [38]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - 604us/step - accuracy: 0.7522 - loss: 0.4981
Loss: 0.49811771512031555, Accuracy: 0.7521865963935852


In [39]:
# Export our model to HDF5 file
nn.save('AlphabetSoupCharity.h5')

So, the required level of `accuracy` `0.7521865963935852` has been achieved, however, let's try to use another optimizer `sgd`, since on some data it shows better results than `adam`.

In [40]:
number_input_features = len(X_train_scaled[0])

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=7, input_dim=number_input_features, activation='relu'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=14, activation='relu'))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=21, activation='tanh'))

# Fourth hidden layer
nn.add(tf.keras.layers.Dense(units=21, activation='linear'))

# Fifth hidden layer
nn.add(tf.keras.layers.Dense(units=28, activation='relu'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()


Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_15 (Dense)                │ (None, 7)              │           525 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_16 (Dense)                │ (None, 14)             │           112 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_17 (Dense)                │ (None, 21)             │           315 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_18 (Dense)                │ (None, 21)             │           462 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_19 (Dense)                │ (None, 28)             │           616 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_20 (Dense)                │ (None, 1)              │            29 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,059 (8.04 KB)

 Trainable params: 2,059 (8.04 KB)

 Non-trainable params: 0 (0.00 B)

In [41]:
nn.compile(loss = 'binary_crossentropy', optimizer = 'sgd', metrics=['accuracy'])

In [42]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,validation_split=0.15, epochs=200)

Epoch 1/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 951us/step - accuracy: 0.5783 - loss: 0.6694 - val_accuracy: 0.6774 - val_loss: 0.6051
Epoch 2/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 787us/step - accuracy: 0.6863 - loss: 0.5846 - val_accuracy: 0.6991 - val_loss: 0.5644
Epoch 3/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 772us/step - accuracy: 0.7010 - loss: 0.5553 - val_accuracy: 0.7064 - val_loss: 0.5456
Epoch 4/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 875us/step - accuracy: 0.7148 - loss: 0.5346 - val_accuracy: 0.7124 - val_loss: 0.5384
Epoch 5/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7193 - loss: 0.5317 - val_accuracy: 0.7142 - val_loss: 0.5291
Epoch 6/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 800us/step - accuracy: 0.7283 - loss: 0.5275 - val_accuracy: 0.7321 - val_loss: 0.5225
Epoch 7/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 770us/step - accuracy: 0.7485 - loss: 0.5110 - val_accuracy: 0.7346 - val_loss: 0.5201
Epoch 8/200
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 773us/step - accuracy: 0.7457 - loss: 0.5

In [43]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - 688us/step - accuracy: 0.7527 - loss: 0.4974
Loss: 0.4973818361759186, Accuracy: 0.7526530623435974


Contrary to classical expectations, `adam` performs worse than `sgd`, with an `Accuracy` of 0.7521865963935852 compared to `sgd`'s `Accuracy` of 0.7526530623435974, and a `Loss` of 0.49811771512031555 compared to `sgd`'s `Loss` of 0.4973818361759186. This is a reason to study the data further.

In [44]:
nn.save('AlphabetSoupCharity_optimized.h5')